### ISSUE TITLE
markdown latex not working in pyodide

### ISSUE THREAD

**Issue by MarcSkovMadsen** [holoviz/panel#8421]
You can reproduce by opening https://panel.holoviz.org/reference/panes/Markdown.html#latex-support and running the example.

<img width="1559" height="474" alt="Image" src="https://github.com/user-attachments/assets/c663e7ff-2a32-470f-9029-78142febbb75" />

I see the same problem when trying to use latex in markdown in pyodide.

**Comment by MarcSkovMadsen** [comment]
# Triage: panel #8421 — Markdown LaTeX not working in Pyodide

**Type:** Bug | **Labels:** `webassembly`, `javascript`

## Summary

LaTeX rendering in the Markdown pane does not work in Pyodide because the `bokeh-mathjax` JS bundle is not included in page resources. Bokeh's `_use_mathjax()` does not recognize Panel's custom `HTML` model as needing MathJax, so `bokeh-mathjax` is excluded from the JS bundle in Pyodide/convert contexts. The `$$...$$` delimited math content is silently not rendered. This also affects `panel convert` output (#4182).

## Next Steps

- [ ] Add a Panel-specific MathJax detection function in `panel/io/resources.py` that checks if any root model is Panel's `HTML` model (note: `disable_math` is NOT on the Bokeh model — see below)
- [ ] Decide: always include `bokeh-mathjax` when any Panel `HTML` model is present (simpler, ~50KB overhead) vs. inspect text content for TeX delimiters (more precise but fragile) vs. threading `disable_math` to the Bokeh model
- [ ] This fix would also resolve [#4182](https://github.com/holoviz/panel/issues/4182) (LaTeX pane not working with `panel convert`)
- [ ] For interactive docs Pyodide context, consider pre-loading `bokeh-mathjax` on reference pages with LaTeX features

<details>
<summary>Details</summary>

**Issue:** [panel#8421](https://github.com/holoviz/panel/issues/8421)
**Author:** @MarcSkovMadsen | **Created:** 2026-02-13 | **State:** open

### Duplicate / Related Issues

- **#4182** (Open): [LaTeX pane not working with panel convert](https://github.com/holoviz/panel/issues/4182) — Same root cause, same reporter. LaTeX pane also fails in `panel convert` output because `bokeh-mathjax` is not loaded.
- **#5340** (Open): [Markdown LaTeX support not showing in docs](https://github.com/holoviz/panel/issues/5340) — Related; LaTeX not rendering in docs examples.
- **#3792** (Open): [Markdown LaTeX rendering issues](https://github.com/holoviz/panel/issues/3792) — General LaTeX rendering issues in Markdown pane.
- **#8416** (Open): [Panel does not work with Pyodide](https://github.com/holoviz/panel/issues/8416) — Broader Pyodide compatibility issues; may be related.

### Documentation Search

The Panel documentation states LaTeX is supported via MathJax when `disable_math=False` (the default) and `pn.extension('mathjax')` is called. However, the documentation does not mention that this may not work in Pyodide/WASM contexts.

### Root Cause

The root cause is in `panel/io/resources.py:bundle_resources()` at line 474. The auto-detection logic for whether to include `bokeh-mathjax` has two paths:

1. **`_use_mathjax(roots)`** — Bokeh's function that inspects model types. It only recognizes Bokeh's own `Div`, `Paragraph`, `TextAnnotation`, `Axis`, `MathText`, etc. It does **not** recognize Panel's custom `HTML` model (from `panel/models/markup.py`), which the Markdown pane uses. Panel's HTML model inherits from Bokeh's `Markup` (not `Div`), so it's invisible to `_use_mathjax()`.

2. **`'mathjax' in ext._loaded_extensions`** — checks if the user explicitly called `pn.extension('mathjax')`. This works when the extension is loaded before resources are bundled, but in the docs interactive Pyodide context, resources are pre-bundled.

The JS rendering pipeline in `panel/models/html.ts:process_tex()` (line 190-213) depends on `this.provider.MathJax` being available. When `bokeh-mathjax` is not loaded, `this.provider.MathJax` is undefined, and `contains_tex()` returns false — LaTeX is silently not rendered.

### Potential Solution

Add a Panel-specific MathJax detection function that checks if any root model is Panel's `HTML` model. This would augment the existing `_use_mathjax(roots)` check.

**Important:** `disable_math` is a Python-level `param.Boolean` on `panel.pane.markup.Markdown` (line 355 of `panel/pane/markup.py`), and is explicitly set to `None` in `_rename` (line 384), meaning it is **NOT synced to the Bokeh model**. The Bokeh `HTML` model (`panel/models/markup.py:26-34`) only has `events` and `run_scripts` properties. Therefore, the detection function cannot check `disable_math` on the Bokeh model. Options:

1. **Simplest:** Always include `bokeh-mathjax` when any Panel `HTML` model is present in roots (recommended — same tradeoff Bokeh makes for its own models)
2. **Alternative:** Thread `disable_math` to the Bokeh model first, then check it
3. **Alternative:** Check at Python level before roots are serialized

```python
def _panel_use_mathjax(roots):
    """Check if any Panel HTML model is present (may need MathJax)."""
    from .models.markup import HTML as PanelHTML
    from bokeh.embed.bundle import _any
    return _any(roots, lambda obj: isinstance(obj, PanelHTML))
```

</details>

<details>
<summary>Reproduction</summary>

**Test file:** `.triaging/triage_panel_8421_20260218_170450_62332c6c-424a-4b2a-9fcd-040b43ceede9_test.py`

**Run tests:**
```bash
python -m pytest .triaging/triage_panel_8421_20260218_170450_62332c6c-424a-4b2a-9fcd-040b43ceede9_test.py -v
```

**Pytest result:** `1 failed, 5 passed in 1.66s`
**Reproduces on main:** Yes (partially)

The core failing test is `test_bundle_resources_includes_mathjax_for_markdown_with_tex` — it confirms that `bokeh-mathjax` is NOT included in JS resources when `bundle_resources()` is called with a Markdown pane containing TeX delimiters.

```python
"""Reproduction tests for panel#8421 — Markdown LaTeX not working in Pyodide.

The issue is that LaTeX rendering in the Markdown pane does not work in Pyodide.
The root cause is that the bokeh-mathjax JS bundle is not included when
resources are bundled for the Pyodide/convert context, because Bokeh's
_use_mathjax() does not recognize Panel's custom HTML model as needing MathJax.

These tests verify:
1. That the Markdown pane correctly preserves TeX delimiters in its HTML output
2. That bundle_resources includes bokeh-mathjax when a Markdown pane with LaTeX is present
3. That bundle_resources includes bokeh-mathjax when 'mathjax' is in loaded extensions
"""
import panel as pn


def _bundle_has_mathjax(bundle):
    """Check if a Bundle object includes bokeh-mathjax in its JS files."""
    js_files_str = " ".join(str(f) for f in bundle.js_files)
    return "mathjax" in js_files_str.lower()


def test_markdown_preserves_tex_delimiters_in_html():
    """Markdown pane should preserve $$...$$ delimiters in the rendered HTML
    so that MathJax can process them on the JS side."""
    md = pn.pane.Markdown(r"$$E = mc^2$$")
    props = md._transform_object(md.object)
    html = props["object"]
    assert "E = mc^2" in html or "E = mc" in html, (
        f"TeX content not found in Markdown HTML output: {html}"
    )


def test_markdown_disable_math_defaults_to_false():
    """Markdown pane should have disable_math=False by default, enabling LaTeX."""
    md = pn.pane.Markdown("test")
    assert md.disable_math is False


def test_bundle_resources_includes_mathjax_for_markdown_with_tex():
    """When a Markdown pane contains TeX, bundle_resources should include
    bokeh-mathjax so that the JS MathJax provider is available.

    This test demonstrates the core bug: Bokeh's _use_mathjax() does not
    detect Panel's HTML model as needing MathJax, so bokeh-mathjax is
    excluded from the JS bundle in Pyodide/convert contexts.
    """
    from bokeh.resources import Resources as BkResources
    from panel.io.resources import Resources, bundle_resources

    md = pn.pane.Markdown(r"$$\int_0^1 x^2 dx = \frac{1}{3}$$")
    doc = md.server_doc()
    roots = doc.roots

    resources = Resources.from_bokeh(BkResources(mode="cdn"), notebook=False)
    bundle = bundle_resources(roots, resources)

    assert _bundle_has_mathjax(bundle), (
        "bokeh-mathjax should be included in JS resources when a Markdown pane "
        "contains TeX delimiters, but it was not found in js_files: "
        f"{[str(f) for f in bundle.js_files]}. "
        "This is the root cause of LaTeX not working in Pyodide."
    )


def test_bundle_resources_includes_mathjax_when_extension_loaded():
    """When 'mathjax' is in the loaded extensions, bundle_resources should
    include bokeh-mathjax regardless of model inspection."""
    from bokeh.resources import Resources as BkResources
    from panel.config import panel_extension
    from panel.io.resources import Resources, bundle_resources

    md = pn.pane.Markdown("No TeX here")
    doc = md.server_doc()
    roots = doc.roots

    resources = Resources.from_bokeh(BkResources(mode="cdn"), notebook=False)

    original = panel_extension._loaded_extensions[:]
    try:
        panel_extension._loaded_extensions = ["mathjax"]
        bundle = bundle_resources(roots, resources)
        assert _bundle_has_mathjax(bundle)
    finally:
        panel_extension._loaded_extensions = original


def test_bundle_resources_includes_mathjax_with_enable_flag():
    """When enable_mathjax=True is passed explicitly, bokeh-mathjax should
    always be included."""
    from bokeh.resources import Resources as BkResources
    from panel.io.resources import Resources, bundle_resources

    md = pn.pane.Markdown("No TeX here")
    doc = md.server_doc()
    roots = doc.roots

    resources = Resources.from_bokeh(BkResources(mode="cdn"), notebook=False)
    bundle = bundle_resources(roots, resources, enable_mathjax=True)
    assert _bundle_has_mathjax(bundle)


def test_convert_script_to_html_includes_mathjax_for_markdown():
    """When converting a script containing Markdown with LaTeX to HTML
    for Pyodide, the output should include bokeh-mathjax."""
    import tempfile
    from panel.io.convert import script_to_html

    script_content = '''
import panel as pn
pn.extension('mathjax')
pn.pane.Markdown(r"$$E = mc^2$$").servable()
'''
    with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
        f.write(script_content)
        f.flush()
        html, _ = script_to_html(
            f.name,
            requirements=[],
            runtime="pyodide",
            prerender=True,
        )

    assert "mathjax" in html.lower()
```

</details>


<details>
<summary>Fix Plan</summary>

### Files to Modify

| File | Change |
|---|---|
| `panel/io/resources.py` | Add `_panel_use_mathjax()` helper that checks if any model in the roots is Panel's `HTML` model. Use it alongside Bokeh's `_use_mathjax()` on line 474. Note: `disable_math` is NOT on the Bokeh model, so the simplest approach is to always include mathjax when any Panel `HTML` model is present. |

### Verification
- [ ] Run reproduction tests: `python -m pytest .triaging/triage_panel_8421_20260218_170450_62332c6c-424a-4b2a-9fcd-040b43ceede9_test.py -v` — all tests pass
- [ ] Run related test suite: `python -m pytest panel/tests/io/test_resources.py -v`
- [ ] Verify in Pyodide: open a Panel Markdown example with LaTeX in the docs and confirm rendering

### Risk Assessment
- Narrowly scoped — only affects the `bundle_resources` function's mathjax detection
- Including `bokeh-mathjax` when not needed adds ~50KB gzipped JS. Same tradeoff Bokeh makes for its own models.
- Since `disable_math=False` is the default, this will include `bokeh-mathjax` whenever a Markdown pane is present. Content-based check is more precise but fragile.
- This fix also addresses #4182 (LaTeX pane not working with panel convert)

</details>

**Comment by etihwo** [comment]
With the following MRE I am not able to reproduce your issue: 

Create an file called `index.py` with the following content
```python
import panel as pn
pn.extension('mathjax')

pn.pane.Markdown(r"""
The Markdown pane supports math rendering for string encapsulated with double $ delimiters: $$\sum_{j}{\sum_{i}{a*w_{j, i}}}$$
""", width=800).servable()
```
Then use the command line `panel convert index.py --out public` and lastly serve the folder with `python -m http.server -d public` And I get : 

<img width="616" height="45" alt="Image" src="https://github.com/user-attachments/assets/c60919c4-aa52-4f11-a576-82d0d8cd71ce" />

<details><summary>Package version</summary>
<p>

|Package    |        Version|
|--------|--------|
| bleach     |         6.3.0 |
| bokeh       |        3.8.2 |
| certifi        |     2026.1.4 |
| charset-normalizer |  3.4.4 |
| colorama     |       0.4.6 |
| contourpy  |        1.3.3 |
| idna             |   3.11 |
| Jinja2          |   3.1.6 |
| linkify-it-py   |    2.0.3 |
| Markdown      |      3.10.2 |
| markdown-it-py  |    4.0.0 |
| MarkupSafe    |      3.0.3 |
| mdit-py-plugins |    0.5.0 |
| mdurl      |         0.1.2 |
| narwhals      |      2.16.0 |
| numpy         |      2.4.2 |
| packaging   |        26.0 |
| pandas       |       3.0.1 |
| panel        |       1.8.7 |
| param        |       2.3.2 |
| pillow      |        12.1.1 |
| pip          |       25.3 |
| python-dateutil  |    2.9.0.post0 |
| pyviz_comms  |       3.0.6 |
| PyYAML      |        6.0.3 |
| requests    |        2.32.5 |
| six             |    1.17.0 |
| tornado    |         6.5.4 |
| tqdm         |       4.67.3 |
| typing_extensions |  4.15.0 |
| tzdata         |     2025.3 |
| uc-micro-py   |      1.0.3 |
| urllib3       |      2.6.3 |
| webencodings    |    0.5.1 |
| xyzservices     |    2025.11.0 |

</p>
</details> 


In addition to the comment above, I took a look on your (LLM generated ?) test and the requirements keyword is different to what will be filled when called by `panel convert`. Please refer to https://github.com/holoviz/panel/blob/b5981ccd0845c40ab6be234078cb0487f5d85284/panel/io/convert.py#L518 When I change the function `test_convert_script_to_html_includes_mathjax_for_markdown` with 

```python
html, _ = script_to_html(
            f.name,
            requirements=['https://cdn.holoviz.org/panel/wheels/bokeh-3.8.2-py3-none-any.whl', 'https://cdn.holoviz.org/panel/1.8.7/dist/wheels/panel-1.8.7-py3-none-any.whl', 'pyodide-http'],
            runtime="pyodide",
            prerender=True,
        )
``` 
I get a valid html file in this function.

**Comment by MarcSkovMadsen** [comment]
Thanks. What do you see when you open https://panel.holoviz.org/reference/panes/Markdown.html#latex-support? Do you see the rendered latex?

**Comment by etihwo** [comment]
In the documentation, when I run the command I have the same issue:

<img width="1030" height="180" alt="Image" src="https://github.com/user-attachments/assets/48126154-612a-4d24-9f5f-4ebd0c05e38e" />

It seems to be only in the documentation since when it is runned in pyodide with `panel convert` the rendering is ok.
